File: JULIA/Cap08/c08ex01.jl
# c08ex01.jl

function caractere_especial(entra::String)::Bool
  caracteres_especiais = "!@#\$%^&*()-_=+[]{}|;:',<>/?`~\""
  for caractere in caracteres_especiais
    if occursin(caractere, entra)
      return true
    end
  end
  return false
end

function valida_inteiro(entra::String)::Bool
  return occursin(r"\A\d+\z", entra)
end

while true
  print("Entre um valor numerico inteiro: ")
  numero = readline()
  println()

  if isempty(numero)
    println("Erro: Tecla <Enter> acionada acidentalmente.\n")
  elseif caractere_especial(numero)
    println("Erro: Caractere especial nao permitido.\n")
  elseif valida_inteiro(numero)
    println("entra valida fornecida: $numero\n")
    break
  else
    println("Erro: Entre um numero inteiro.\n")
  end
end

print("\nTecle <Enter> para encerrar... ")
readline()

File: RUST/Cap08/c08ex01.rs
// c08ex01.rs

use std::io::{self, Write};

fn caractere_especial(entra: &str) -> bool {
  let caractere_especial = "!@#$%^&*()-_=+[]{}|;:',<>/?`~\"";
  for caractere in caractere_especial.chars() {
    if entra.contains(caractere) {
      return true;
    }
  }
  false
}

fn valida_inteiro(entra: &str) -> bool {
  entra.parse::<i32>().is_ok()
}

fn main() {
  loop {
    print!("Entre um valor numerico inteiro: ");
    io::stdout().flush().expect("Falha ao limpar o buffer");

    let mut numero = String::new();
    io::stdin().read_line(&mut numero).expect("Falha ao ler a entrada");
    numero = numero.trim().to_string();

    println!();

    if numero.is_empty() {
      println!("Erro: Tecla <Enter> acionada acidentalmente.\n");
    } else if caractere_especial(&numero) {
      println!("Erro: Caractere especial nao permitido.\n");
    } else if valida_inteiro(&numero) {
      println!("Entrada valida fornecida: {}\n", numero);
      break;
    } else {
      println!("Erro: Entre um numero inteiro.\n");
    }
  }

  println!();
  print!("Tecle <Enter> para encerrar... ");
  io::stdout().flush().expect("Falha ao limpar o buffer");
  let mut _entrada = String::new();
  io::stdin().read_line(&mut _entrada).expect("Falha ao ler a entrada");
}
